### ISSUE TITLE
Non-ASCII characters in test files lead to parsing errors on Windows

### ISSUE THREAD

**Issue by mbojan** [r-lib/testthat#1589]
This seems to be related to #1496 . I'm getting parsing errors on Windows when `test_file()`ing files that include non-ASCII characters, but which are in UTF-8. The attached file [the-test.zip](https://github.com/r-lib/testthat/files/8253486/the-test.zip) is an R script, just rename to extension `.R`. Running `enca -L polish` on that file will tell that it is indeed in UTF-8.

I'm getting:

```
testthat::test_file("the-test.R")

== Testing the-test.R ==============================================================================================
[ FAIL 0 | WARN 0 | SKIP 0 | PASS 0 ]Error in parse(con, n = -1, srcfile = srcfile, encoding = "UTF-8") : 
  the-test.R:1:24: unexpected $end
1: d <- tibble::tibble(ImiÄ™
                           ^
```

so clearly some encoding issues. Digging further points to parsing:

```
> traceback()
12: parse(con, n = -1, srcfile = srcfile, encoding = "UTF-8")
11: source_file(path, child_env(env), wrap = wrap)
10: FUN(X[[i]], ...)
9: lapply(test_paths, test_one_file, env = env, wrap = wrap)
8: doTryCatch(return(expr), name, parentenv, handler)
7: tryCatchOne(expr, names, parentenv, handlers[[1L]])
6: tryCatchList(expr, classes, parentenv, handlers)
5: tryCatch(code, testthat_abort_reporter = function(cnd) {
       cat(conditionMessage(cnd), "\n")
    ...
4: with_reporter(reporters$multi, lapply(test_paths, test_one_file, 
       env = env, wrap = wrap))
3: test_files(test_dir = test_dir, test_package = test_package, 
       test_paths = test_paths, load_helpers = load_helpers, reporter = reporter, 
    ...
2: test_files(test_dir = dirname(path), test_package = package, 
       test_paths = basename(path), reporter = reporter, ...)
1: testthat::test_file("the-test.R")
```

Interstingly the test file seems to be read correctly by this line (did some tracing and inspected the result)

https://github.com/r-lib/testthat/blob/22a68dd85b65c95173424760a78955b6c55f3848/R/source.R#L18

but the call to `parse()` two steps later fails, even though `encoding` is set to UTF-8 everywhere (on the connection and for `parse()`).

<details>
<summary>Session info</summary>

      - Session info ----------------------------------------------------------------------------------------------------
       setting  value
       version  R version 4.1.3 (2022-03-10)
       os       Windows 10 x64 (build 19042)
       system   x86_64, mingw32
       ui       RStudio
       language (EN)
       collate  Polish_Poland.1250
       ctype    Polish_Poland.1250
       tz       Europe/Warsaw
       date     2022-03-15
       rstudio  2021.09.0+351 Ghost Orchid (desktop)
       pandoc   2.17.1.1 @ C:\\PROGRA~3\\CHOCOL~1\\bin\\pandoc.exe
      
      - Packages --------------------------------------------------------------------------------------------------------
       package     * version    date (UTC) lib source
       brio          1.1.3      2021-11-30 [1] CRAN (R 4.1.2)
       cli           3.2.0      2022-02-14 [1] CRAN (R 4.1.2)
       crayon        1.5.0      2022-02-14 [1] CRAN (R 4.1.2)
       desc          1.4.1      2022-03-06 [1] CRAN (R 4.1.2)
       ellipsis      0.3.2      2021-04-29 [1] CRAN (R 4.1.0)
       fansi         1.0.2      2022-01-14 [1] CRAN (R 4.1.2)
       glue          1.6.2      2022-02-24 [1] CRAN (R 4.1.2)
       lifecycle     1.0.1      2021-09-24 [1] CRAN (R 4.1.1)
       magrittr      2.0.2      2022-01-26 [1] CRAN (R 4.1.2)
       pillar        1.7.0      2022-02-01 [1] CRAN (R 4.1.2)
       pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 4.1.0)
       pkgload       1.2.4      2021-11-30 [1] CRAN (R 4.1.1)
       R6            2.5.1      2021-08-19 [1] CRAN (R 4.1.1)
       rlang         1.0.2      2022-03-04 [1] CRAN (R 4.1.2)
       rprojroot     2.0.2      2020-11-15 [1] CRAN (R 4.1.0)
       rsthemes      0.2.1.9000 2021-10-13 [1] Github (gadenbuie/rsthemes@19299e5)
       rstudioapi    0.13       2020-11-12 [1] CRAN (R 4.1.0)
       sessioninfo   1.2.2      2021-12-06 [1] CRAN (R 4.1.2)
       testthat    * 3.1.3      2022-03-15 [1] Github (r-lib/testthat@22a68dd)
       tibble        3.1.6      2021-11-07 [1] CRAN (R 4.1.1)
       utf8          1.2.2      2021-07-24 [1] CRAN (R 4.1.0)
       vctrs         0.3.8      2021-04-29 [1] CRAN (R 4.1.0)
       withr         2.5.0      2022-03-03 [1] CRAN (R 4.1.2)
      
       [1] C:/Users/mbojanowski/Documents/R/library/x86_64-w64-mingw32/4.1
       [2] C:/Users/mbojanowski/Documents/Programs/R-4.1.3/library
      
      -------------------------------------------------------------------------------------------------------------------

</details>

**Comment by hadley** [comment]
What version of R are you using?

**Comment by mbojan** [comment]
This was on 4.1.3, see the collapsed session info at the bottom of issue description

![obraz](https://user-images.githubusercontent.com/636574/191508035-8b440783-1bc6-4c47-9878-e30f8caa1721.png)

**Comment by hadley** [comment]
Is it resolved by R 4.2?

**Comment by mbojan** [comment]
Yep. Works OK on 4.2.1.

**Comment by hadley** [comment]
So do you mind if I close this issue? It's probably possible to fix in R 4.1, but it's likely to be a few hours work investigation for me, so if using R 4.2 solves the problem for you, I'd rather prioritise something else.

**Comment by hadley** [comment]
Yeah, one of the big features in 4.2 is radically improved handling of UTF-8 on windows.
